txn.go

```diff
@@ -0,0 +1,155 @@
+package mongods
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	dsextensions "github.com/textileio/go-datastore-extensions"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+var (
+	ErrTxnFinalized = errors.New("txn was already finalized")
+)
+
+type mongoTxn struct {
+	// lock serializes all API access since the
+	// mongo session isn't goroutine-safe as mentioned
+	// in the docs.
+	lock      sync.Mutex
+	finalized bool
+
+	m       *MongoDS
+	session mongo.Session
+	ctx     mongo.SessionContext
+}
+
+var _ datastore.Txn = (*mongoTxn)(nil)
+
+func (m *MongoDS) NewTransaction(readOnly bool) (datastore.Txn, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	if m.closed {
+		return nil, ErrClosed
+	}
+
+	session, err := m.m.StartSession()
+	if err != nil {
+		return nil, fmt.Errorf("starting mongo session: %s", err)
+	}
+
+	if err := session.StartTransaction(); err != nil {
```

Comment:
We start a session and a transaction within that session.
The session reference is kept in the `txn` so operations work inside the txn.